### ISSUE TITLE
Pinning first column and highlighting row

### ISSUE THREAD

**Issue by proxseas** [man-group/dtale#559]
I've been using this library more and more and I was wondering if it was possible to have a few more "quality of life" improvements. I'll list a few of them here, instead of creating multiple threads.

1. It would be super useful to be able to 'pin' the first column, so that when you keep scrolling far to the right, you'll still be able to see that important first column. In Excel and Google Sheets, this is called 'freezing'.
2. Highlight a row when you click on it - just as in Excel or Google Sheets. Relatedly, I think that highlighting a row upon mouse hover wouldn't be useful.
3. Ability to further filter an already filtered set - when you do so using by clicking the filter options in the UI. For example, you click on a column, then click on '>', then enter a number 'x' - and now you've filtered the data by values greater than 'x'. Now, you have a couple buttons that appeared above the grid. One of them is '(x)' - for clearing all filters. Another is the eraser mark, for dropping the filtered rows. What I think would be very helpful here is:
- The ability to hide the 'drop rows' button that looks like an eraser. I've gotten used to avoiding it, but I think it would be useful if it was possible to hide it if you don't ever need to drop rows.
- An 'edit' button that takes the current query, e.g. `numcol > 2.3` and simply opens the 'custom filter' dialog with this query pre-populated. This would allow extending the current filtering very seamlessly (without having to copy & paste the currently active filter query into the input box of the Custom Filters modal). Also, I realize I can just press Shift+F and this would open up the Custom Filters dialog, as well, but since there are existing buttons, why not add one for this purpose as well?

4. Ability to color rows in a more customized way. What I would really like to do is color them based on a certain value range. The closest thing to this is 'Highlight Range', which looks like it's almost exactly what I'd need, but it's currently limited to the color yellow. Ideally, it would be possible to set up more than 1 condition (e.g. if greater than x, then color it red, if greater than y, then color it orange, if less than z, color it blue). Personally, I don't need a gradient for my purposes, but I imagine some people might.
Additionally, it would be very meaningful if it was possible to configure this in code, before launching dtale. Otherwise, I'd need to set it up manually every time I run dtale.
I'd like to point out the obvious and say that this is also Excel-like functionality. But it really makes the data easier to make sense of at a quick glance.

5. This is more of a question, than a feature request - I could not figure out how to set it up in code (prior to launching dtale) such that 'NaN' values are rendered as '-'. In the UI, it's possible to set this up under "Display "nan" values as"..
I've been able to set up other numeral.js column formats, just not this one.

All this said, thank you so much for this great library. It really makes pandas so much more satisfying to use!

**Comment by aschonfeld** [comment]
@proxseas just wanted to give you a heads up that I'll starting working on this shortly.  In the meantime, your first request for pinning columns already exists but it's called "locking" instead of "pinning".  Right now, if your input dataframe has an index then those columns in the index will be "pinned" to the left of the grid but you can also pin any other columns by clicking the header and pressing the "Lock" button as shown here:
![image](https://user-images.githubusercontent.com/11547371/132391437-cfe60f9a-8015-4ab6-9439-6e2e50290cf0.png)

And to unlock simply click the header of the column you just locked and press the "Unlock" button.

**Comment by proxseas** [comment]
That's very convenient. I should have paid more attention to the drop down menu. Is it possible to pre-set a specific column programatically (that isn't the index) to lock before launching dtale?

**Comment by aschonfeld** [comment]
@proxseas not sure if this is what you are talking about for (2) but you can already select rows and [copy them into excel](https://github.com/man-group/dtale#copy-cells-into-clipboard).  That article shows you how to copy cells.  In order to copy rows simply Shift+Click on the row indexes on the left.

**Comment by proxseas** [comment]
@aschonfeld I was not trying to copy and paste into Excel. I was only using it for referencing functionality.

**Comment by aschonfeld** [comment]
@proxseas I was able to update clicking of row indexes so that it will highlight that specific row for you: 

https://user-images.githubusercontent.com/11547371/132958611-7076a17e-1861-4e38-9220-8e8711567050.mov

**Comment by proxseas** [comment]
Looking good!

**Comment by aschonfeld** [comment]
@proxseas I'm starting on (3) and what I'm planning to do is make the "drop rows" button configurable from the the config file or python code.  Similar to how you can turn off cell editing by setting a flag to `False`.  This will be an application-scoped flag so that if you set it to `False` it will be turned off for all instances of D-Tale.

So for the config file you would add something like:
```
[app]
drop_rows = False
```

Or from code you would do:
```
import dtale

dtale.global_state.set_app_settings(dict(drop_rows=False))
```

For reference, here's more information on [configuration](https://github.com/man-group/dtale/blob/master/docs/CONFIGURATION.md) (apologies if I've sent this before).

The "edit" button which moves the column-based filters into the custom filters is doable and a very clever idea.

**Comment by aschonfeld** [comment]
Here's what I got for (3): 

https://user-images.githubusercontent.com/11547371/133182158-958e8927-2bf2-459d-b1ae-d3f911277588.mov

**Comment by proxseas** [comment]
Can't wait for the release.

**Comment by aschonfeld** [comment]
For (5) in order to set the "Nan Display" you would have to do the following:

```
import dtale
import dtale.global_state as global_state
import pandas as pd

df = pd.DataFrame([1,2,3,4,5])
d = dtale.show(df)
settings = global_state.get_settings(d._data_id) or {}
settings["nanDisplay"] = "..."
global_state.set_settings(d._data_id, settings)
```

This is a little roundabout and I'll work on adding a parameter to `dtale.show` named `nan_display` which can take any value and the front-end will inject it into any `nan` cells.

**Comment by aschonfeld** [comment]
For (5) you can currently achieve this using the following hack:
```
import dtale
import dtale.global_state as global_state
import pandas as pd
import numpy as np

df = pd.DataFrame(dict(a=[1, 2, 3, np.nan, 5]))
d = dtale.show(df)
settings = global_state.get_settings(d._data_id) or {}
settings["nan_display"] = "..."
global_state.set_settings(d._data_id, settings)
```

I plan on consolidating this so that you can do something like `dtale.show(df, nan_display="...")`.  Also, nan_display will only work for `"nan", "inf", "-", ""` in numeric columns but any string (EX: `"..."`) will work for strings or dates.  I will be fixing this as well.

**Comment by aschonfeld** [comment]
FYI, I'm moving (4) to a new ticket since its a larger scale change.  Definitely doable, but I have a couple other requests I need to take care of before then.

**Comment by aschonfeld** [comment]
These features (with the exception of (4)) have been released in v1.57.0
